src/lib/HST/Effect/PatternStack.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE TemplateHaskell, ScopedTypeVariables #-}
+{-# LANGUAGE LambdaCase, BlockArguments #-}
+
+-- | This module defines an effect for computations that can save patterns
+--   variables have been matched against on a stack.
+
+module HST.Effect.PatternStack
+  ( -- * Effect
+    PatternStack
+    -- * Actions
+  , pushPattern
+  , peekPattern
+  , popPattern
+    -- * Interpretations
+  , runPatternStack
+  )
+where
+
+import           Data.Map                       ( Map )
+import qualified Data.Map                      as Map
+import           Polysemy                       ( Sem
+                                                , makeSem
+                                                , reinterpret
+                                                )
+import           Polysemy.State                 ( State
+                                                , evalState
+                                                , gets
+                                                , modify
+                                                )
+
+
+import qualified HST.Frontend.Syntax           as S
+
+-------------------------------------------------------------------------------
+-- Effect and Actions                                                        --
+-------------------------------------------------------------------------------
+
+-- | An effect capable of pushing and poping patterns variables have been
+--   matched against to a stack.
+data PatternStack a m b where
+  PushPattern ::S.QName a -> S.Pat a -> PatternStack a m ()
+  PeekPattern ::S.QName a -> PatternStack a m (Maybe (S.Pat a))
+  PopPattern ::S.QName a -> PatternStack a m ()
+
+makeSem ''PatternStack
+
+-------------------------------------------------------------------------------
+-- Interpretations                                                           --
+-------------------------------------------------------------------------------
+
+-- | A map of stacks of patterns by variable names.
+type StackMap a = Map (S.QName a) [S.Pat a]
+
+-- | Interprets the given computation by providing one stack for each variable.
+runPatternStack :: Sem (PatternStack a ': r) b -> Sem r b
+runPatternStack = evalState Map.empty . patternStackToState
+ where
+  patternStackToState
+    :: Sem (PatternStack a ': r) b -> Sem (State (StackMap a) ': r) b
+  patternStackToState = reinterpret \case
+    PushPattern name pat -> modify $ Map.alter (maybeCons pat) name
+    PeekPattern name     -> gets $ fmap head . Map.lookup name
```

Comment:
The computation cannot fail since entries are added and removed with `PushPattern` and `PopPattern` only. `PopPattern` removes the entry from the map if the list is empty via `maybeTail`. Thus, `Map.lookup` (and consequently `PeekPattern`) returns `Nothing` if there is no pattern for `name`. If `Map.lookup` returns a `Just`, the list of patterns is guaranteed to be non-empty.